content/advent-2016/advanced-encoding-decoding.md

```diff
@@ -0,0 +1,507 @@
+
+The way these two interfaces work is pretty straight forward; When the `encoding/json` package runs into a type that implements the `Marshaler` interface, it uses that type's `MarshalJSON()` method instead of the default marshaling code to turn the object into JSON. Similarly, when decoding a JSON object it will test to see if the object implements the `Unmarshaler` interface, and if so it will use the `UnmarshalJSON()` method instead of the default unmarshaling behavior. That means that all we need to do to ensure our `Dog` type is always encoded and decoded as a `JSONDog` is to implement these two methods and do the conversion there.
+
+Once again, we are going to start with encoding first, which means we will be implementing the `MarshalJSON() ([]byte, error)` method on our `Dog` type. 
+
+While this might feel like a large undertaking at first, we can actually utilize a lot of our existing code to minimize what we need to write. All we really need to do in this method is return the results from calling `json.Marshal()` on a `JSONDog` representation of our current `Dog`.
+
+```go
+func (d Dog) MarshalJSON() ([]byte, error) {
+  return json.Marshal(NewJSONDog(d))
+}
+```
+
+Now it doesn't matter if a developer forgets to convert our `Dog` type to the `JSONDog` type; This will happen by default when the `Dog` is encoded into JSON.
+
+The `Unmarshaler` implementation ends up being pretty similar. We are going to implement the `UnmarshalJSON([]byte) error` method, and once again we are going to utilize our existing `JSONDog` type.
+
+```go
+func (d *Dog) UnmarshalJSON(data []byte) error {
+  var jd JSONDog
+  if err := json.Unmarshal(data, &jd); err != nil {
+    return err
+  }
+  *d = jd.Dog()
+  return nil
+}
+```
+
+Finally, we update our `main()` function to use the `Dog` type for both decoding and encoding instead of the `JSONDog` type.
+
+```go
+func main() {
+  dog := Dog{1, "bowser", "husky", time.Now()}
+  b, err := json.Marshal(dog)
+  if err != nil {
+    panic(err)
+  }
+  fmt.Println(string(b))
+
+  b = []byte(`{
+    "id":1,
+    "name":"bowser",
+    "breed":"husky",
+    "born_at":1480979203}`)
+  dog = Dog{}
+  json.Unmarshal(b, &dog)
+  fmt.Println(dog)
+}
+```
+
+*You can find a working sample of the code up until this point on the Go Playground here: <https://play.golang.org/p/GR6ckydMxF>*
+
+With about 10 lines of code we have overridden the default JSON encoding for our `Dog` type. Pretty neat, right?
+
+Next we are tackle the other problem we had with our initial approach - we are going to use aliases to reduce the amount of repeated code we need to use.
+
+
+
+### Using embedded aliases to reduce code
```

Comment:
I opted to put a little disclaimer here in the hopes that it helps prevent confusion. I'm still open to suggestions on better terms, as I'm not 100% sure what the correct terminology should be.